[v3] Prepare for v3

ErikSchierboom opened the pull request:
_This issue is part of the migration to v3. You can read full details about the various changes [here](https://github.com/exercism/v3-launch/)._

This PR prepares the track for Exercism v3, which will be different in a number of ways from Exercism v2.

In this PR, the following changes are made:

## v2 file migration

1. Move the existing exercises in the `exercises` directory to the `exercises/practice` directory

## v3 file migration

1. Copy the v3 `languages/<slug>/concepts` directory to the `concepts` directory
1. Copy the v3 `languages/<slug>/exercises/concept` directory to the `exercises/concept` directory
1. Copy the v3 `languages/<slug>/reference` directory to the `reference` directory
1. Copy the v3 `languages/<slug>/docs` directory to the `docs` directory

### Notes

Any commits modifying the aforementioned directories (or any of their files) are included in the PR, except for:

- The `_sidebar.md` files are not migrated
- The `README.md` file at the v3 track's root is not migrated.
- The `maintainers.md` file is not migrated.

## config.json migration

1. Add a version property:

```json
"version": 3
```

2. Add online editor settings:

```json
"online_editor": {
  "indent_style": "space",
  "indent_size": 2
}
```

3. Rename the `"exercises"` property to `"practice"`

4. Create an `"exercises"` property and move the `"practice"` property to this property:

```json
"exercises": {
  "practice": [
    ...
  ]
}
```

5. Remove the `"core"`, `"auto_approve"`, and `"unlocked_by"` properties from the practice exercises

6. Add the `"name"` property to the practice exercises and pre-populate this with a titlelized version of the `"slug"` property

7. Add the `"prerequisites"` property to the practice exercises and set it to an empty array

8. Add an empty `"concept"` array property to the `"exercises"` property

```json
"exercises": {
  "concept": [],
  "practice": [
    ...
  ]
}
```

9. Move the `"foregone"` property to the `"exercises"` key:

```json
"exercises": {
  ...
  "foregone": [...]
}
```

10. Add a top-level `"concepts"` key, which is an array:

```json
"concepts": []
```

11. Add a top-level `"tags"` key, which is an array:

```json
"tags": []
```

12. Add a top-level `"key_features"` key, which is an array:

```json
"key_features": []
```

13. Add a top-level `"v3_ready"` key, which is a boolean value that is set to `true` if the track has a `config.json` file in the v3 repo, and otherwise set to `false`:

```json
"v3_ready": true
```

14. Add a top-level `"slug"` key, which is a string containing the track's slug:

```json
"slug": "csharp"
```

### Notes

- Settings defined in the v3 `config.json` file will take precedence over their v2 `config.json` file's equivalent.

## configlet

1. Update the `fetch-configlet` and `fetch-configlet.ps1` files to the [latest version of configlet (v3)](https://github.com/exercism/canonical-data-syncer), which can work with v3 tracks.

## Continuous integration

1. Add a GitHub Actions workflow to verify the track using the [configlet-CI GitHub Action](https://github.com/exercism/github-actions/tree/master/configlet-ci), unless there already is a file named `.github/workflows/configlet.md`.

1. Add a [dependabot configuration](https://dependabot.com/github-actions/) to automatically submit PRs for any new versions of external workflows used in this track's GitHub Action workflows.

## Follow-up steps

We've created issues in this repo for the follow-up steps to get this track ready for v3.

ErikSchierboom committed:
Add docs/reference split

* [Docs] Move implementing-a-concept-exercise.md to reference folder
* [Docs] Move reference documents to docs folder
* [Docs] Update to student-facing docs
* [Docs] Add new issue template

Co-Authored-By: Jeremy Walker <jez.walker@gmail.com>
Co-Authored-By: Victor Goff <keeperotphones@gmail.com>
Co-authored-by: Sascha Mann <git@mail.saschamann.eu>

yawpitch committed:
Concept extraction / references mega merge

* Create common working area

* Extract Concepts from v2 exercise: reverse-string

* Create reverse-string.md

* Update reverse-string.md

* Add Concepts from v2 exercise: variable-length-quantity

* Add first concepts group

* Improved concepts as per PR review

* Adds concept from binary-search-tree

* Add initial list

First pass concepts for `allergies` to address #460

* Initial list of concepts

First pass list of concepts to address #459

* Add Concepts for v2 exercise: phone-number

* Add phone-number Python concepts

* Small update to index access and slice topics.

* Add notes from review.
 - more information about classes, inheritance
 - flesh out privacy, public and non-public
 - clarify wording around iterables and index/slice access

* One more note about brackets and strings.

* Add Concepts for v2 exercise: hamming

* Add concepts for hamming

* Add note about tuple unpacking.

* Add notes about polymorphism, builtins, and dunder methods.

* Some whitespace fixes.

* [WIP] `clock` exercise concepts.

* Extract Concepts from v2 exercise: markdown

* Initial commit for markdown exercise concepts.

* Concept starter for markdown

* Added detail to Markdown concepts

* Final edits before harmonization

Final Markdown edits before we merge and harmonize.

* Add Concepts for v2 exercise: matrix

* `matrix` exercise concepts (issue #386)

First pass of concepts for `matrix ` exercise in python.  Pretty sure this is too detailed, but wanted to get something for review before proceeding with additional exercises.

* Edits to better match #290 Formatting

Edited concepts to better match the formatting of issue #290

* Typo correction

* added title

* Extract Concepts from v2 exercise: rna-transcription

* Beginning of Concepts for rna-transcription

* More detailed concepts for rna-trranscription

More detailed concepts for rna-transcription exrcise.

* Added title

* Extract Concepts from v2 exercise: robot-simulator

* Beginning of concepts for robot-simulator.

* WIP Concepts

* Additional detail for concepts

* Detail third pass

Third pass on adding concept detail.

* Additional detail for concepts.

* Edits per PR Feedback

Numerous spelling corrections.  Additional edits to address comments from last review.

* [WIP] Concept implementation instructions

* Adds instructions for exercise implementation

* Adds correction as per PR reviews

* Harmonize, part 1

* fix relative links in references/README.md

* First pass at harmonization

Shifts all documents to a common format, adds minimal link tagging to
the "concept" currently listed in each file. These will really need
multiple more passes, as they diverge from each other even when
describing the same topic. Many extraneous topics have crept in, added
in an "aspirational" fashion to the exercises; we may need to trim some
of that.

* Pulling in examples from BethanyG

* [WIP] Extracted concept unification

* Unification of extracted concepts

* Typos and duplicates remove

* Duplicates concept unification

* Concepts have now links to original file

* Update languages/reference/README.md

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

Co-authored-by: khoivan88 <33493502+khoivan88@users.noreply.github.com>
Co-authored-by: David G <davidgerva@gmail.com>
Co-authored-by: Ashley Drake <a.l.drake713@gmail.com>
Co-authored-by: Pedro Romano <pedro@paparomeo.net>
Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

yawpitch committed:
Organize concepts, eliminate duplication

* Stage 1

* Finish restructuring, add checkboxes.

* Minor format fixes

* fixing broken link

* fixing broken link

* fixing README link

* Fix multiples bad links

* Fix links in reference/concepts/keywords/README.md

* Correcting list description

aldraco committed:
Port concept exercise strings from C#

* Add scaffold files/directories.

* Add instructions and introduction.

* Add hints with some links.

* Add config files.

* Update description to use loglines module, and add prerequisite.

* Add Python specific files, start on implementation.

* Update tests.

* Add more tests.

* Update docs with text around links.

* Update tests with reformat test, and fix example import from testing.

* Add third function, reformat.

* Move an item into after.md, and add Joel Spolsky's blog post as reference for strings.

* Add design notes.

* Add reference types note.

* Alternate solution not using regex groups, just split.

* Update languages/exercises/concept/strings/.docs/after.md

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Update loglines test
 - updated filename to match tooling
 - import functions, not module
 - use testcase.assertEqual instead of assert

* Use an example that does not use Regex.

* Better regex for regex-based solution.

* Update .docs and .meta.

* Change strings-methods to strings-basic, for consistency and to better reflect the purpose of the exercise.

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Add more information to introduction.md. TODO: add links.

* Add extra context for hints.md.

* Add UUID to strings exercise config.

* Simpify initial description of strings.

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Add more context around various string methods.

* Remove After header.

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Use link for Joel on Software post.

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Cleanup header formatting, use actual link.

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Change method -> function.

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Name prerequisite with -basic suffix.

* Leave a bit to the imagination for strings hints.

* Move example.py into .meta folder per instructions.

* Update links for after and hints.md

* Update hints.md w.r.t. join() and fix link.

* Add missing implementation link.

* Single spacing between functions.

* Remove leading whitespace in

* Fix missing backticks and make Python examples read as if inside a REPL.

* Update from review feedback.

* Update languages/exercises/concept/strings/.docs/hints.md

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/strings/.docs/instructions.md

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/strings/.docs/introduction.md

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/strings/.docs/hints.md

Co-Authored-By: Erik Schierboom <erik_schierboom@hotmail.com>

* Link to python3 latest docs.

* Use named capture groups in regex for solution.

* Add extra test for capitalized words.

* Use msg parameter for assert statements in test.

* Use dashed style UUID.

* Rename test and clarify exactly what is being tested.

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

ErikSchierboom committed:
Add reference to required reading in implementing guide

[Docs] Add reference to required reading in implementing guide

yawpitch committed:
Update .meta/config.json

Updates as per changes to the file schema: fixes #1192.

ErikSchierboom committed:
Add reference to Concept Exercise Anatomy video

[Docs] Add reference to Concept Exercise Anatomy video

ErikSchierboom committed:
Format using Prettier

[Docs] Format using Prettier

yawpitch committed:
Renames loglines*.py -> strings*.py

The existing naming would conflict with the expectations of the
test runner / representer / analyzer tooling.

ErikSchierboom committed:
Cross-reference concept exercise file information

[Docs] Cross-reference concept exercise file information

DavidGerva committed:
exercises table update

* new concept exercises issues

* new exiting loops concept exercise issue

* Fix rendering issue

* applied prettier

* Add exiting loops reference page

* Improve exiting-loops concept

* add reference

* fix links

* fix broken links

ErikSchierboom committed:
Use consistent headings in documentation

ErikSchierboom committed:
Remove top-level header from design.md files

morrme committed:
rearrange functions to match instructions

Fixes #1850

BethanyG committed:
Add tuples concept exercise

* New tuples concept exercise design file.

* New tuples concept exercise example implementation file.

* New tuples concept exercise config  file.

* New tuples concept exercist after file w TODO.

* New tuples concept exercist hints file.

* New tuples concept exercise introduction file.

* New tuples concept exercise instructions file.

* New tuples concept exercise boilerplate file.

* New tuples concept exercise cleaned boilerplate file.

* New tuples concept exercise tests.

* Ran prettier to clean up file markdown formatting.  Edited language and cleaned up bloat.

* Ran prettier to correct markdown formatting and dleted star operator as learning objective.

* Removed tuple multiplication with star operator.

* Edited to remove multiply function.

* Edited to remove multiply function test case.

* Fixed quoting on Scrimshaw Whale's tooth to match quoting style.

* Ran prettier for config.json

* Edits and adjustments per PR feedback and style guides.  Also corrected introduction.md file name spelling.

* Edited for clarity and brevity per PR feedback.  Removed excess concepts and code examples.

* Re-worked final task to be a returned report string rather than print()-ing of content.  Adjusted instructions accordingly.

* Cleaned up per PR feedback.  Removed excess lines.

* Changed final test to support altered task for report string.  Removed context handler decorator for re-routing stdout and extra imports.

* Corrected terminal output code exxample for task 4.

* Removed print as a prerequisite of the exercise.  Added absolute link for tuples.md refrence doc.

* Removed excess PyTest config from test class.

* Added tuples concept exercise information and uuid.

* Ran prettier on config file, since I forgot to.

BethanyG committed:
Ran Prettier on instructions.md for Lists Concept exercise.

* Initial draft of "lists" concept exercise for Python track (Issue #859)

* Ran prettier to format markdown instructions.md file.

Co-authored-by: Itamar Gal <itamar@blueorange.digital>

InfiniteVerma committed:
specify function names

* specify function names

* [CI] Format code

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

neenjaw committed:
new concept exercise - `bools`

Introduces the bools concept, see design and .docs for more details.

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

BethanyG committed:
new concept exercise - `basics`

* Introduction to Python basics concept exercise.

* Intructions to Python basics concept exercise.

* Hints file for Python basics concept exercise.

* After file for Python basics concept exercise.

* Design and config files for Python basics exercise.

* Example implementation of Python basics concept exercise.

* Stub file and tests for Python basics concept exercise.

* Ran prittier. AGAIN.

* Update languages/exercises/concept/basics/.docs/after.md

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

* Update languages/exercises/concept/basics/.docs/hints.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/basics/lasagna.py

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/basics/.docs/introduction.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/basics/.docs/introduction.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/basics/.docs/after.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Language and link edits per PR suggestions by iHiD, ErikSchierboom, and cmccandless.

* Added basics exercise slug and UUID.

* Ran Prettier.

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

J08K committed:
Implement new Concept Exercise: dicts (new PR)

* Added it again

* Updated the following to the requested changes.

* Fixed some spelling in after.md

* Add stuff to after.md

* formatting

* Some touches here and there and added example.py

* Create test

* Update dicts_test.py

* formatting

* Update after.md

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/example.py

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/dicts_test.py

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/hints.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/after.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update after.md

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/design.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/design.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/design.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/design.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/design.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update design.md

* Update languages/exercises/concept/dicts/.meta/example.py

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/example.py

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/example.py

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.meta/example.py

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Rewriting introduction.md and adding some stuff to after.md

* Update example.py

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update introduction.md

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/dicts/.docs/instructions.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update instructions.md

* Rework

* Update introduction.md

* Update languages/exercises/concept/dicts/.meta/example.py

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>
Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

ErikSchierboom committed:
Add stubs for concept documents

mohanrajanr committed:
Python Concept Exercise enums

Pre concept exercise re-format.

mohanrajanr committed:
Implement new Concept Exercise: list-methods #2170

* Added New Concept Exercise List Methods Part 1

* Review Change 1

mohanrajanr committed:
Implement new Concept Exercise: loops #1430

* Loops Concept Exercise First Change

* Review Change 1

* Update languages/exercises/concept/loops/.docs/instructions.md

* Update languages/exercises/concept/loops/.docs/instructions.md

* Update languages/exercises/concept/loops/.docs/instructions.md

* Update languages/exercises/concept/loops/.docs/instructions.md

* Update languages/exercises/concept/loops/.docs/instructions.md

* Update languages/exercises/concept/loops/.docs/instructions.md

* Update languages/exercises/concept/loops/.docs/instructions.md

* Update languages/exercises/concept/loops/.docs/introduction.md

* Update languages/exercises/concept/loops/.docs/introduction.md

* Update languages/exercises/concept/loops/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

mohanrajanr committed:
Implement new Concept Exercise: none #1031

* Added Python Concept Exercise None

* Ran Prettier for Linting

* Updated Story and new Exercises

* Review 1 None-Intro.md

* Review Changes 2

ErikSchierboom committed:
bools - replace about.md file with concept files

* Pre-populate bools concept's about.md file from after.md file

* Pre-populate bools concept's links.json file from after.md file

* bools - Remove after.md document

* Update languages/concepts/bools/links.json

* Update languages/concepts/bools/links.json

* Update languages/concepts/bools/links.json

* Update languages/concepts/bools/links.json

* Update languages/concepts/bools/links.json

* Update languages/concepts/bools/links.json

* Update languages/concepts/bools/links.json

* [CI] Format code

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

ErikSchierboom committed:
tuples - replace about.md file with concept files

* Pre-populate tuples concept's about.md file from after.md file

* Pre-populate tuples concept's links.json file from after.md file

* tuples - Remove after.md document

* Apply suggestions from code review

Added link descriptions.

* Update languages/concepts/tuples/links.json

Re-ordered and edited links.

* [CI] Format code

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

ErikSchierboom committed:
basics - replace about.md file with concept files

ErikSchierboom committed:
strings - replace about.md file with concept files

* Pre-populate strings concept's about.md file from after.md file

* Pre-populate strings concept's links.json file from after.md file

* strings - Remove after.md document

* Apply suggestions from code review

Removed exercise specific info and added more detail on the concept.  Added and re-arranged links.

* [CI] Format code

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

BethanyG committed:
Bools: Add Links and Clean Up Concept Doc

* Added concept links and additional information  to bools concept.

* Edited intro to refer to bool type not boolean.

* Removed built-in types requirements and added basics and comparison operators.

* [CI] Format code

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

BethanyG committed:
Concept Exercises Rename and Configuration

* Renamed list-methods exercise.

* Renamed lists exercise.

* Renamed bools exercise.

* Renamed basics exercise.

* Renamed dicts exercise.

* Renamed Enums exercise.

* Renamed the loops exercise.

* Renamed the strings exercise.

* Renamed the none exercise.

* Copied and renamed tuples exercise as starter and placeholder for sets exercise.

* Renamed tuples exercise.

* Deletions of old file names for concept exercises.

* Big exercise rename and requirements re-do.

* Updated some track statuses.

* Fixed lasagna typo.

* Corrected broken link to python strings exercise.

* Added renamed concept exercises to story files to fix broken links.

* More broken links fixed for pac-man and strings types.

* Update reference/stories/enums.log-line-parser.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Update languages/config.json

* Renamed elyses-enchantment-exercises to elyses-enchantments and updated config.json.

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

angelikatyborska committed:
Update mix app names; Rename freelancer-rate -> freelancer-rates

Co-authored-by: Tim Austin <tim@neenjaw.com>

ErikSchierboom committed:
Add concept headings to introduction.md

BethanyG committed:
Basics Concept Exercise Reformat - A Re-implementation of PR 2534

* Edited basics exercise and change inline links to refrence links.

* Alphabetized in-doc links, edited down introduction.md and edited down links.json

* More editing down of introduction.md

ErikSchierboom committed:
Add description of concept documents

[Docs] Add description of concept documents

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

valentin-p committed:
Inventory Dicts Improvement

* Inventory Dicts Improvment

* proofing the about file

* check working

* add Bethany

* Update languages/concepts/dicts/about.md

* Update languages/concepts/dicts/about.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

BethanyG committed:
Changed in doc links to ref links in markdown. Edited and fixed code examples.

BethanyG committed:
Added numbers concept and links.json.

Yabby1997 committed:
Fix typo in line 27, from defind to define

valentin-p committed:
enums. iterate on the log-level exercise

* enums. iterate on the log-level exercise

* chnage hints and other files

* minor changes

* minor fixes

* straightforward changes

* [CI] Format code

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

BethanyG committed:
List-Method Exercise Edits and List Concept Doc

* Added lists concept about doc.

* Edited instructions and introduction docs.

* Ran prittier.

* Update languages/concepts/lists/about.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

* Update languages/concepts/lists/about.md

* Update languages/concepts/lists/about.md

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/chaitanas-colossal-coaster/.docs/introduction.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/concepts/lists/about.md

* [CI] Format code

* [CI] Format code

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>
Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

kimolivia committed:
Implement new concept exercise - string methods #2175

* Add new concept exercise - string methods

* Fix formatting

* Use prettier to clean up markdown

* Resolve PR comments

* Fix closing bracket

* PR comment, add concept files

* Update languages/exercises/concept/string-methods/.meta/design.md

* Update languages/concepts/string-methods/about.md

* Update languages/concepts/string-methods/about.md

* Update languages/concepts/string-methods/about.md

* Update languages/exercises/concept/string-methods/.docs/introduction.md

Co-authored-by: Olivia Kim <oliviakim@Olivias-MacBook-Pro.local>
Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

BethanyG committed:
Rename string-methods Concept Exercise to "Little Sisters Essay"

* Renamed concept exercise to story name.

* Exercise rename to story title.

* added one link and trailing newline.

* Added uuid and entry for str-methods (little sisters essay) exercise.

* Fixed typo in story title.

OMEGA-Y committed:
Fix typo and remove white space in about.md

valentin-p committed:
dicts improve links

* dicts improve links

* revert to two lines. Surround top-level function and class definitions with two blank lines.

valentin-p committed:
str-methods. Fix https prefix and typo

valentin-p committed:
list methods. code format and proofread

* change python code format and proofread

BethanyG committed:
Replace then with them

spotted typo.

serenalucca committed:
Improve Concept Exercise: lists #2880

* Improve Concept Exercise: lists #2880

Hi! This is my very first contribution to an open-source project, I need feedback, especially for the hints file.
Thanks in advance.

* Improve Concept Exercise: lists #2880

applied prettier

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.meta/design.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* slice notation in introduction and about file

* modification of hint and introduction file + format of comments

* prettier format

* Update hints.md

* Update languages/concepts/lists/about.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update languages/concepts/lists/about.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* Update about.md

* Update introduction.md

* prettier

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

SaschaMann committed:
Add missing concepts files

* Add missing concepts files

* [CI] Format code

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

Limm-jk committed:
Add new concept exercise : numbers

* First

* Add : config.json
authors & editor

* Add : exchange file
exchange_test / exchange

* Fix : typo

Line 2
from .exchange import * -> from exchange import *

* Fix : Forder Name

* Fix : Folder Name

* Added example and instructions.

added : currency-exchange/.meta/example.py
added : currency-exchange/.docs/instructions.md

* Added design.md

* added design.md
added : currency-exchange/.meta/design.md

* Added hint.md

Added hint.md

* Add : hint.md

* Delete .DS_Store

* Add : hint.md

Add : hint.md
Solving Conflict

* Add introduction.md

* Fix: typo

fix exchangable to exchangeable
fix "Chandler want" to "Chandler wants" in instructions.md

* Change File_name & Prettier

* Fix : Prettier

* Add : slug to config.json

* Fix config

* Fix introduction.md

* Fix instructions.md

* Fixed instructions.md

Fixed grammatical errors in instructions.md based on BethanyG's review.

Co-authored-by: Limm-jk <201602057@cs-cnu.org>
Co-authored-by: Junkyu Lim <57378834+Limm-jk@users.noreply.github.com>

* add testcase

* Added problem "6. Calculate unexchangeable value"

* Added new problem "Calculate unexchangeable value"

Added modular arithmetic problem Calculate unexchangeable value

* Added new problem "Calculate unexchangeable value" (example.py, exchange.py)

Modified example.py, exchange.py

* Modified Calculated unexchangeable value

Modified `unexchangeable_value()` function's expected return type to be `int`

* added new lines at the end of file

* applied prettier on instructions.md

Co-authored-by: Limm-jk <201602057@cs-cnu.org>
Co-authored-by: Junkyu Lim <57378834+Limm-jk@users.noreply.github.com>

* Add : Problem Design

* Update languages/config.json

we changed the concept name, so needed to update file.

* Update languages/exercises/concept/currency-exchange/.docs/introduction.md

* Update languages/exercises/concept/currency-exchange/.docs/introduction.md

* Update languages/exercises/concept/currency-exchange/.docs/introduction.md

* Update languages/exercises/concept/currency-exchange/.docs/introduction.md

* Update languages/exercises/concept/currency-exchange/.meta/config.json

we've decided to name the stub and test files for the concept and not the story.

* Update languages/exercises/concept/currency-exchange/.meta/design.md

minor typo

* Update languages/exercises/concept/currency-exchange/exchange_test.py

* Update languages/exercises/concept/currency-exchange/exchange_test.py

* Update languages/exercises/concept/currency-exchange/.docs/introduction.md

Co-authored-by: Yunseon Choi <science_co@naver.com>

* Update languages/config.json

removed arithmetic as a concept until we get the about.md file made.

* Update languages/exercises/concept/currency-exchange/.meta/config.json

reverting file name change and deferring to different PR.  This is currently breaking a CI test.

Co-authored-by: Yoggy <wjdek88@naver.com>
Co-authored-by: Seunghun Yang <Yabby1997@gmail.com>
Co-authored-by: Ticktakto <50654571+Ticktakto@users.noreply.github.com>
Co-authored-by: Yunseon Choi <science_co@naver.com>
Co-authored-by: wnstj2007 <43427152+wnstj2007@users.noreply.github.com>
Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

BethanyG committed:
Concepts rename pt2

* renamed stub files and adjusted config file.  Corrected reference to floating point numbers and added additonal info on precision.

* renamed stub files and corrected config file.

* Edited code commet formatting and adjusted content.

* Added missing uuids and changed concept names for strings.

* Changed the name of str related concepts and numbers related concepts.  Updated numbers exercise info.

* Added additional links.  Renamed a few concepts.

* renamed stub files and adjusted config file.  Corrected reference to floating point numbers and added additonal info on precision.

* renamed stub files and corrected config file.

* Edited code commet formatting and adjusted content.

* Added missing uuids and changed concept names for strings.

* Changed the name of str related concepts and numbers related concepts.  Updated numbers exercise info.

* Added additional links.  Renamed a few concepts.

* ran prettier. Again.

* Ran prettier for currency exchange.

* Updated concept WIP table with numbers exercise and reanmed heirarchical concepts.

* Update languages/exercises/README.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

* Update languages/exercises/README.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

OMEGA-Y committed:
Fix typo design.md

OMEGA-Y committed:
Fix markdown

OMEGA-Y committed:
Fix typo

OMEGA-Y committed:
add link design.md

BethanyG committed:
Updates to implementing-a-concept-exercise.md

* Updated Python concept exercise implementation doc.

* [CI] Format code

* Updated Python concept exercise implementation doc.

* added backticks around concepts we don't want writers to use unless already covered.

* re-worded example.py file explaination.

* And yet, I had to run prettier again.

* Update languages/reference/implementing-a-concept-exercise.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

* Update languages/reference/implementing-a-concept-exercise.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

* Update languages/reference/implementing-a-concept-exercise.md

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

BethanyG committed:
Concepts naming cleanup and typo patrol

* Another edit pass for missing links, typoes, and explainations.

* Cleanup, typos, added new examples and comments.

* Cleanup, typos, formatting.

* Cleanup typos, float explaination re-word.

* Changed class concept naming.

* Brought inro file more in lie with concept about file.

* Brought intro file more in line with concept about file.

* Changed the naming of the class concepts.

* Update languages/concepts/lists/about.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

* Update languages/exercises/concept/elyses-enchantments/.docs/introduction.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

ErikSchierboom committed:
Cleanup instructions headings

ErikSchierboom committed:
Add concept introductions

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

valentin-p committed:
Review of the lists elyses concept exercise

* changing arg names
improving the code formatting

* review about file

* remove comparisons from the config

* Apply suggestions from code review

* [CI] Format code

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

valentin-p committed:
string formatting concept exercise

* config and first draft for pretty leaflet

* finishing the exercise and documentation

* reviews from Bethany

* add more details about i18n and safe

* add msg for each string

* Update languages/concepts/string-formatting/about.md

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

* add words

Co-authored-by: BethanyG <BethanyG@users.noreply.github.com>

BethanyG commented:
Did a quick double check, and I think things are "good to go".  *thank you* @ErikSchierboom for all of your work on this!

ErikSchierboom commented:
Great!

ErikSchierboom committed:
Convert example to exemplar

* Rename example files to exemplar

See https://github.com/exercism/v3-docs/pull/23

* [Docs] Correct .meta/example references to .meta/exemplar

* [Docs] Use exemplar instead of example

* [Docs] Update example name in file listings

* [Julia] Convert to exemplar.jl

* [elm] Rename .meta/Cook.elm to .meta/Examplar.elm

* [elm] Rename .meta/Examplar.elm .meta/Exemplar.elm

Co-authored-by: Matthieu Pizenberg <matthieu.pizenberg@gmail.com>

sachsom95 committed:
Implement new Concept Exercise: conditionals #1103

* Created all required files for python concept - conditionals

* [CI] Format code

* CUsed prettier to format the files

* Used prettier to format the files

* First draft of introduction.md done

* Initial draft completed on introduction.md and working on after.md

* Made first draft of after.md

* Update languages/exercises/concept/conditionals/.docs/after.md

David cleaned up Grammar issues

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

minor grammar issues fixed from davids suggestion

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

grammar corrections ahh sigh!!

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: David G <davidgerva@gmail.com>

* Modifications to after.md as per review
-  The modifications made at "one line if statement"
- The inital draft had mentioned about one line if statement after the example for it
-  Now the one line if statement style guide is before showing example

* Minor corrections as stated in initial review
- Chnaged the eerenious output at line 34 for elif statement

* Created the initial draft for design.md

* Modified 'config.json'
- Misinterpreted "forked-from" now removed that and ammeded the mistake

* Update languages/exercises/concept/conditionals/.docs/introduction.md

Accepting Grammar corrections from Eric

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

* Cleaned up and refined the after.md for approval

The after.md has been edited and cleaned. The file seems to be
easy to port to new docs for the concept exercises

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>

* Update languages/exercises/concept/conditionals/.docs/after.md

* Created the instruction.md
A draft of instruction.md is made with a theme for review
Refactoring will be required later to compile everything to new folder structure

* Update languages/exercises/concept/conditionals/.docs/instructions.md

Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

* Second Iteration
- Removed Most of the Typos
- Made the formating more in line with standars used
- Grmmar and words changed

* added the functions definitions and the parameters

* Generated the scaffolding for test cases

* Test cases for first function added to conditionals_test.py
Minor corrections to instruction.py

* Added test cases for second function in conditional_test.py
- Minor logical corrections in instruction.py

* Added the final functions test cases

* [CI] Format code

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: David G <davidgerva@gmail.com>
Co-authored-by: Corey McCandless <cmccandless@users.noreply.github.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>
Co-authored-by: valentin-p <3833193+valentin-p@users.noreply.github.com>

cmccandless committed:
Update CI test runner for v3 (#2284)

* - Check for exemplar.py (>=v3), then example.py (<v3)
- Use pathlib instead of os.path

* remove f-strings for 3.5 compat

* Revert "remove f-strings for 3.5 compat"

This reverts commit c72e129dd2a50e629de0db2cefaa1794be713dba.

* Remove 3.5 from test matrix, add 3.9

* run workflow on main branch too

ErikSchierboom committed:
Convert editor key in .meta/config.json files

* Convert editor key to files

* Add missing "files" field to .meta/config.json files

* Add exemplar files to .meta/config.json

* Add common-lisp exemplar to template

ErikSchierboom committed:
[v3] Move existing exercises to exercises/practice

ErikSchierboom committed:
[v3] Add status to config.json

ErikSchierboom committed:
[v3] Add version property to config.json

ErikSchierboom committed:
[v3] Add online_editor property to config.json

ErikSchierboom committed:
[v3] Add slug to config.json

ErikSchierboom committed:
[v3] Re-order practice exercises in config.json

ErikSchierboom committed:
[v3] Add status for deprecated practice exercises in config.json

ErikSchierboom committed:
[v3] Add (empty) prerequisites property to practice exercises in config.json

ErikSchierboom committed:
[v3] Remove deprecated properties from practice exercises in config.json

ErikSchierboom committed:
[v3] Add name property to practice exercises in config.json

ErikSchierboom committed:
[v3] Move exercises to practice exercises property in config.json

ErikSchierboom committed:
[v3] Add concept exercises to config.json

ErikSchierboom committed:
[v3] Add wip status to concept exercises in config.json

ErikSchierboom committed:
[v3] Move foregone exercises to exercises property in config.json

ErikSchierboom committed:
[v3] Add concepts to config.json

ErikSchierboom committed:
[v3] Add configlet CI workflow

ErikSchierboom committed:
[v3] Add tags to config.json

ErikSchierboom committed:
[v3] Add key features to config.json

ErikSchierboom committed:
[v3] Update fetch-configlet script to work with configlet v3

ErikSchierboom committed:
[v3] Use main branch for Exercism GitHub Actions

ErikSchierboom committed:
[v3] Convert relative v3 links to absolute links

ErikSchierboom committed:
[v3] Add dependabot to keep GHA dependencies up-to-date

ErikSchierboom committed:
[v3] Fix configlet CI workflow

cmccandless committed:
Remove configlet step from ci-workflow

configlet has its own workflow now

cmccandless committed:
fix syntax error

cmccandless committed:
fix syntax errors

cmccandless committed:
remove unused incomplete function

cmccandless committed:
temporarily disable failing checks
